File: go/pkg/doubaospeech/protocol.go
package doubaospeech

import (
	"bytes"
	"compress/gzip"
	"encoding/binary"
	"fmt"
	"io"
)

// ================== 协议常量 ==================

type protocolVersion byte
type messageType byte
type messageTypeFlags byte
type serializationType byte
type compressionType byte

const (
	protocolVersionV1 protocolVersion = 0b0001

	// Message Types
	msgTypeFullClient      messageType = 0b0001
	msgTypeAudioOnlyClient messageType = 0b0010
	msgTypeFullServer      messageType = 0b1001
	msgTypeAudioOnlyServer messageType = 0b1011
	msgTypeFrontEndResult  messageType = 0b1100
	msgTypeError           messageType = 0b1111

	// Message Type Specific Flags
	msgFlagNoSequence  messageTypeFlags = 0b0000
	msgFlagPosSequence messageTypeFlags = 0b0001
	msgFlagNegSequence messageTypeFlags = 0b0010
	msgFlagNegWithSeq  messageTypeFlags = 0b0011
	msgFlagWithEvent   messageTypeFlags = 0b0100

	// Serialization Types
	serializationNone   serializationType = 0b0000
	serializationJSON   serializationType = 0b0001
	serializationThrift serializationType = 0b0011

	// Compression Types
	compressionNone compressionType = 0b0000
	compressionGzip compressionType = 0b0001

	// Protocol Event Types
	eventSessionStart       int32 = 1
	eventSessionFinish      int32 = 2
	eventConnectionStarted  int32 = 50
	eventConnectionFailed   int32 = 51
	eventConnectionFinished int32 = 52
)

// ================== 协议结构 ==================

// binaryProtocol 二进制协议处理器
//
// 协议格式:
// - Header (4 bytes):
//   - (4bits) version + (4bits) header_size
//   - (4bits) message_type + (4bits) message_type_flags
//   - (4bits) serialization + (4bits) compression
//   - (8bits) reserved
//
// - Payload:
//   - [optional] sequence (4 bytes)
//   - [optional] event (4 bytes)
//   - [optional] session_id (4 bytes len + data)
//   - payload_size (4 bytes) + payload_data
type binaryProtocol struct {
	version       protocolVersion
	headerSize    byte
	compression   compressionType
	serialization serializationType
}

// message 协议消息
type message struct {
	msgType   messageType
	flags     messageTypeFlags
	event     int32
	sessionID string
	connectID string
	sequence  int32
	errorCode uint32
	payload   []byte
}

// newBinaryProtocol 创建协议处理器
func newBinaryProtocol() *binaryProtocol {
	return &binaryProtocol{
		version:       protocolVersionV1,
		headerSize:    1, // 4 bytes
		compression:   compressionNone,
		serialization: serializationJSON,
	}
}

// setCompression 设置压缩方式
func (p *binaryProtocol) setCompression(c compressionType) {
	p.compression = c
}

// setSerialization 设置序列化方式
func (p *binaryProtocol) setSerialization(s serializationType) {
	p.serialization = s
}

// marshal 序列化消息
func (p *binaryProtocol) marshal(msg *message) ([]byte, error) {
	buf := new(bytes.Buffer)

	// Header (4 bytes)
	buf.WriteByte(byte(p.version<<4) | p.headerSize)
	buf.WriteByte(byte(msg.msgType<<4) | byte(msg.flags))
	buf.WriteByte(byte(p.serialization<<4) | byte(p.compression))
	buf.WriteByte(0x00) // reserved

	// Sequence (if needed)
	if msg.flags&msgFlagPosSequence != 0 || msg.flags&msgFlagNegSequence != 0 {
		if err := binary.Write(buf, binary.BigEndian, msg.sequence); err != nil {
			return nil, fmt.Errorf("write sequence: %w", err)
		}
	}

	// Event (if needed)
	if msg.flags&msgFlagWithEvent != 0 {
		if err := binary.Write(buf, binary.BigEndian, msg.event); err != nil {
			return nil, fmt.Errorf("write event: %w", err)
		}

		// Session ID (for non-connection events)
		if msg.event != eventSessionStart && msg.event != eventSessionFinish &&
			msg.event != eventConnectionStarted &&
			msg.event != eventConnectionFailed &&
			msg.event != eventConnectionFinished {
			if err := binary.Write(buf, binary.BigEndian, uint32(len(msg.sessionID))); err != nil {
				return nil, fmt.Errorf("write session id length: %w", err)
			}
			buf.WriteString(msg.sessionID)
		}
	}

	// Payload
	payload := msg.payload
	if p.compression == compressionGzip && len(payload) > 0 {
		compressed, err := gzipCompress(payload)
		if err != nil {
			return nil, fmt.Errorf("gzip compress: %w", err)
		}
		payload = compressed
	}

	if err := binary.Write(buf, binary.BigEndian, uint32(len(payload))); err != nil {
		return nil, fmt.Errorf("write payload size: %w", err)
	}
	buf.Write(payload)

	return buf.Bytes(), nil
}

// unmarshal 反序列化消息
func (p *binaryProtocol) unmarshal(data []byte) (*message, error) {
	if len(data) < 4 {
		return nil, fmt.Errorf("data too short: %d bytes", len(data))
	}

	buf := bytes.NewBuffer(data)

	// Read header
	versionAndSize, _ := buf.ReadByte()
	typeAndFlags, _ := buf.ReadByte()
	serAndComp, _ := buf.ReadByte()
	_, _ = buf.ReadByte() // reserved

	msg := &message{
		msgType: messageType(typeAndFlags >> 4),
		flags:   messageTypeFlags(typeAndFlags & 0x0f),
	}

	compression := compressionType(serAndComp & 0x0f)

	// Header size (in 4-byte units)
	headerSize := int(versionAndSize & 0x0f)
	if headerSize > 1 {
		// Skip additional header bytes
		buf.Next((headerSize - 1) * 4)
	}

	// Read sequence if present
	if msg.flags&msgFlagPosSequence != 0 || msg.flags&msgFlagNegSequence != 0 {
		if err := binary.Read(buf, binary.BigEndian, &msg.sequence); err != nil {
			return nil, fmt.Errorf("read sequence: %w", err)
		}
	}

	// Read event if present
	if msg.flags&msgFlagWithEvent != 0 {
		if err := binary.Read(buf, binary.BigEndian, &msg.event); err != nil {
			return nil, fmt.Errorf("read event: %w", err)
		}

		// Read session ID (for non-connection events)
		if msg.event != eventSessionStart && msg.event != eventSessionFinish &&
			msg.event != eventConnectionStarted &&
			msg.event != eventConnectionFailed &&
			msg.event != eventConnectionFinished {
			var sessionIDLen uint32
			if err := binary.Read(buf, binary.BigEndian, &sessionIDLen); err != nil {
				return nil, fmt.Errorf("read session id length: %w", err)
			}
			if sessionIDLen > 0 {
				sessionIDBytes := make([]byte, sessionIDLen)
				if _, err := buf.Read(sessionIDBytes); err != nil {
					return nil, fmt.Errorf("read session id: %w", err)
				}
				msg.sessionID = string(sessionIDBytes)
			}
		}

		// Read connect ID for connection events
		if msg.event == eventConnectionStarted ||
			msg.event == eventConnectionFailed ||
			msg.event == eventConnectionFinished {
			var connectIDLen uint32
			if err := binary.Read(buf, binary.BigEndian, &connectIDLen); err != nil {
				return nil, fmt.Errorf("read connect id length: %w", err)
			}
			if connectIDLen > 0 {
				connectIDBytes := make([]byte, connectIDLen)
				if _, err := buf.Read(connectIDBytes); err != nil {
					return nil, fmt.Errorf("read connect id: %w", err)
				}
				msg.connectID = string(connectIDBytes)
			}
		}
	}

	// Read error code for error messages
	if msg.msgType == msgTypeError {
		if err := binary.Read(buf, binary.BigEndian, &msg.errorCode); err != nil {
			return nil, fmt.Errorf("read error code: %w", err)
		}
	}

	// Read payload
	var payloadSize uint32
	if err := binary.Read(buf, binary.BigEndian, &payloadSize); err != nil {
		return nil, fmt.Errorf("read payload size: %w", err)
	}

	if payloadSize > 0 {
		msg.payload = make([]byte, payloadSize)
		if _, err := buf.Read(msg.payload); err != nil {
			return nil, fmt.Errorf("read payload: %w", err)
		}

		// Decompress if needed
		if compression == compressionGzip {
			decompressed, err := gzipDecompress(msg.payload)
			if err != nil {
				return nil, fmt.Errorf("gzip decompress: %w", err)
			}
			msg.payload = decompressed
		}
	}

	return msg, nil
}

// isAudioOnly 是否为纯音频消息
func (msg *message) isAudioOnly() bool {
	return msg.msgType == msgTypeAudioOnlyServer || msg.msgType == msgTypeAudioOnlyClient
}

// isError 是否为错误消息
func (msg *message) isError() bool {
	return msg.msgType == msgTypeError
}

// isFrontend 是否为前端结果消息
func (msg *message) isFrontend() bool {
	return msg.msgType == msgTypeFrontEndResult
}

// gzipCompress gzip 压缩
func gzipCompress(data []byte) ([]byte, error) {
	var buf bytes.Buffer
	w := gzip.NewWriter(&buf)
	if _, err := w.Write(data); err != nil {
		return nil, err
	}
	if err := w.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// gzipDecompress gzip 解压
func gzipDecompress(data []byte) ([]byte, error) {
	r, err := gzip.NewReader(bytes.NewBuffer(data))
	if err != nil {
		return nil, err
	}
	defer r.Close()
	return io.ReadAll(r)
}

File: rust/doubaospeech/src/protocol.rs
//! Binary protocol for Realtime and Podcast V3 services.
//!
//! Protocol format:
//! - Header (4 bytes):
//!   - (4bits) version + (4bits) header_size
//!   - (4bits) message_type + (4bits) message_type_flags
//!   - (4bits) serialization + (4bits) compression
//!   - (8bits) reserved
//!
//! - Payload:
//!   - [optional] sequence (4 bytes)
//!   - [optional] event (4 bytes)
//!   - [optional] session_id (4 bytes len + data)
//!   - payload_size (4 bytes) + payload_data

use std::io::{Cursor, Read, Write};

use bytes::{Buf, BufMut, BytesMut};
use flate2::read::GzDecoder;
use flate2::write::GzEncoder;
use flate2::Compression;

use crate::error::{Error, Result};

// ================== Protocol Constants ==================

/// Protocol version.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum ProtocolVersion {
    #[default]
    V1 = 0b0001,
}

/// Message type.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum MessageType {
    #[default]
    FullClient = 0b0001,
    AudioOnlyClient = 0b0010,
    FullServer = 0b1001,
    AudioOnlyServer = 0b1011,
    FrontEndResult = 0b1100,
    Error = 0b1111,
}

impl From<u8> for MessageType {
    fn from(v: u8) -> Self {
        match v {
            0b0001 => MessageType::FullClient,
            0b0010 => MessageType::AudioOnlyClient,
            0b1001 => MessageType::FullServer,
            0b1011 => MessageType::AudioOnlyServer,
            0b1100 => MessageType::FrontEndResult,
            0b1111 => MessageType::Error,
            _ => MessageType::FullClient,
        }
    }
}

/// Message type flags.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum MessageFlags {
    #[default]
    NoSequence = 0b0000,
    PosSequence = 0b0001,
    NegSequence = 0b0010,
    NegWithSeq = 0b0011,
    WithEvent = 0b0100,
}

impl From<u8> for MessageFlags {
    fn from(v: u8) -> Self {
        match v {
            0b0000 => MessageFlags::NoSequence,
            0b0001 => MessageFlags::PosSequence,
            0b0010 => MessageFlags::NegSequence,
            0b0011 => MessageFlags::NegWithSeq,
            0b0100 => MessageFlags::WithEvent,
            _ => MessageFlags::NoSequence,
        }
    }
}

/// Serialization type.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum SerializationType {
    #[default]
    None = 0b0000,
    Json = 0b0001,
    Thrift = 0b0011,
}

impl From<u8> for SerializationType {
    fn from(v: u8) -> Self {
        match v {
            0b0000 => SerializationType::None,
            0b0001 => SerializationType::Json,
            0b0011 => SerializationType::Thrift,
            _ => SerializationType::None,
        }
    }
}

/// Compression type.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum CompressionType {
    #[default]
    None = 0b0000,
    Gzip = 0b0001,
}

impl From<u8> for CompressionType {
    fn from(v: u8) -> Self {
        match v {
            0b0000 => CompressionType::None,
            0b0001 => CompressionType::Gzip,
            _ => CompressionType::None,
        }
    }
}

/// Protocol event types.
pub mod events {
    pub const SESSION_START: i32 = 1;
    pub const SESSION_FINISH: i32 = 2;
    pub const CONNECTION_STARTED: i32 = 50;
    pub const CONNECTION_FAILED: i32 = 51;
    pub const CONNECTION_FINISHED: i32 = 52;
    
    // Realtime events
    pub const ASR_STARTED: i32 = 100;
    pub const ASR_FINISHED: i32 = 101;
    pub const TTS_STARTED: i32 = 102;
    pub const TTS_FINISHED: i32 = 103;
    pub const AUDIO_RECEIVED: i32 = 104;
}

// ================== Protocol Message ==================

/// Binary protocol message.
#[derive(Debug, Clone, Default)]
pub struct Message {
    /// Message type.
    pub msg_type: MessageType,
    /// Message flags.
    pub flags: MessageFlags,
    /// Event type (if flags include WithEvent).
    pub event: i32,
    /// Session ID.
    pub session_id: String,
    /// Connect ID.
    pub connect_id: String,
    /// Sequence number.
    pub sequence: i32,
    /// Error code (for error messages).
    pub error_code: u32,
    /// Payload data.
    pub payload: Vec<u8>,
}

impl Message {
    /// Creates a new message.
    pub fn new() -> Self {
        Self::default()
    }

    /// Creates an audio-only client message.
    pub fn audio_only(session_id: &str, audio: Vec<u8>, event: i32) -> Self {
        Self {
            msg_type: MessageType::AudioOnlyClient,
            flags: MessageFlags::WithEvent,
            event,
            session_id: session_id.to_string(),
            payload: audio,
            ..Default::default()
        }
    }

    /// Returns true if this is an audio-only message.
    pub fn is_audio_only(&self) -> bool {
        matches!(
            self.msg_type,
            MessageType::AudioOnlyServer | MessageType::AudioOnlyClient
        )
    }

    /// Returns true if this is an error message.
    pub fn is_error(&self) -> bool {
        self.msg_type == MessageType::Error
    }

    /// Returns true if this is a frontend result message.
    pub fn is_frontend(&self) -> bool {
        self.msg_type == MessageType::FrontEndResult
    }

    /// Returns true if flags include WithEvent.
    pub fn has_event(&self) -> bool {
        self.flags == MessageFlags::WithEvent
    }
}

// ================== Binary Protocol ==================

/// Binary protocol encoder/decoder.
#[derive(Debug, Clone)]
pub struct BinaryProtocol {
    version: ProtocolVersion,
    header_size: u8,
    compression: CompressionType,
    serialization: SerializationType,
}

impl Default for BinaryProtocol {
    fn default() -> Self {
        Self::new()
    }
}

impl BinaryProtocol {
    /// Creates a new binary protocol handler.
    pub fn new() -> Self {
        Self {
            version: ProtocolVersion::V1,
            header_size: 1, // 4 bytes
            compression: CompressionType::None,
            serialization: SerializationType::Json,
        }
    }

    /// Sets compression type.
    pub fn set_compression(&mut self, compression: CompressionType) {
        self.compression = compression;
    }

    /// Sets serialization type.
    pub fn set_serialization(&mut self, serialization: SerializationType) {
        self.serialization = serialization;
    }

    /// Marshals a message to bytes.
    pub fn marshal(&self, msg: &Message) -> Result<Vec<u8>> {
        let mut buf = BytesMut::with_capacity(256);

        // Header (4 bytes)
        buf.put_u8((self.version as u8) << 4 | self.header_size);
        buf.put_u8((msg.msg_type as u8) << 4 | (msg.flags as u8));
        buf.put_u8((self.serialization as u8) << 4 | (self.compression as u8));
        buf.put_u8(0x00); // reserved

        // Sequence (if needed)
        if msg.flags == MessageFlags::PosSequence || msg.flags == MessageFlags::NegSequence {
            buf.put_i32(msg.sequence);
        }

        // Event (if needed)
        if msg.flags == MessageFlags::WithEvent {
            buf.put_i32(msg.event);

            // Session ID (for non-connection events)
            if !is_connection_event(msg.event) {
                buf.put_u32(msg.session_id.len() as u32);
                buf.put_slice(msg.session_id.as_bytes());
            }
        }

        // Payload
        let payload = if self.compression == CompressionType::Gzip && !msg.payload.is_empty() {
            gzip_compress(&msg.payload)?
        } else {
            msg.payload.clone()
        };

        buf.put_u32(payload.len() as u32);
        buf.put_slice(&payload);

        Ok(buf.to_vec())
    }

    /// Unmarshals bytes to a message.
    pub fn unmarshal(&self, data: &[u8]) -> Result<Message> {
        if data.len() < 4 {
            return Err(Error::Other(format!(
                "data too short: {} bytes",
                data.len()
            )));
        }

        let mut cursor = Cursor::new(data);

        // Read header
        let version_and_size = cursor.get_u8();
        let type_and_flags = cursor.get_u8();
        let ser_and_comp = cursor.get_u8();
        let _reserved = cursor.get_u8();

        let mut msg = Message {
            msg_type: MessageType::from(type_and_flags >> 4),
            flags: MessageFlags::from(type_and_flags & 0x0f),
            ..Default::default()
        };

        let compression = CompressionType::from(ser_and_comp & 0x0f);

        // Header size (in 4-byte units)
        let header_size = (version_and_size & 0x0f) as usize;
        if header_size > 1 {
            // Skip additional header bytes
            cursor.advance((header_size - 1) * 4);
        }

        // Read sequence if present
        if msg.flags == MessageFlags::PosSequence || msg.flags == MessageFlags::NegSequence {
            msg.sequence = cursor.get_i32();
        }

        // Read event if present
        if msg.flags == MessageFlags::WithEvent {
            msg.event = cursor.get_i32();

            // Read session ID (for non-connection events)
            if !is_connection_event(msg.event) {
                let session_id_len = cursor.get_u32() as usize;
                if session_id_len > 0 {
                    let mut session_id_bytes = vec![0u8; session_id_len];
                    cursor.copy_to_slice(&mut session_id_bytes);
                    msg.session_id = String::from_utf8_lossy(&session_id_bytes).to_string();
                }
            }

            // Read connect ID for connection events
            if is_connection_event(msg.event) {
                let connect_id_len = cursor.get_u32() as usize;
                if connect_id_len > 0 {
                    let mut connect_id_bytes = vec![0u8; connect_id_len];
                    cursor.copy_to_slice(&mut connect_id_bytes);
                    msg.connect_id = String::from_utf8_lossy(&connect_id_bytes).to_string();
                }
            }
        }

        // Read error code for error messages
        if msg.msg_type == MessageType::Error {
            msg.error_code = cursor.get_u32();
        }

        // Read payload
        let payload_size = cursor.get_u32() as usize;
        if payload_size > 0 {
            let mut payload = vec![0u8; payload_size];
            cursor.copy_to_slice(&mut payload);

            // Decompress if needed
            if compression == CompressionType::Gzip {
                payload = gzip_decompress(&payload)?;
            }

            msg.payload = payload;
        }

        Ok(msg)
    }
}

/// Returns true if the event is a connection-level event.
fn is_connection_event(event: i32) -> bool {
    matches!(
        event,
        events::SESSION_START
            | events::SESSION_FINISH
            | events::CONNECTION_STARTED
            | events::CONNECTION_FAILED
            | events::CONNECTION_FINISHED
    )
}

/// Gzip compress data.
fn gzip_compress(data: &[u8]) -> Result<Vec<u8>> {
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder
        .write_all(data)
        .map_err(|e| Error::Other(format!("gzip compress: {}", e)))?;
    encoder
        .finish()
        .map_err(|e| Error::Other(format!("gzip finish: {}", e)))
}

/// Gzip decompress data.
fn gzip_decompress(data: &[u8]) -> Result<Vec<u8>> {
    let mut decoder = GzDecoder::new(data);
    let mut decompressed = Vec::new();
    decoder
        .read_to_end(&mut decompressed)
        .map_err(|e| Error::Other(format!("gzip decompress: {}", e)))?;
    Ok(decompressed)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_marshal_unmarshal() {
        let proto = BinaryProtocol::new();

        let msg = Message {
            msg_type: MessageType::AudioOnlyClient,
            flags: MessageFlags::WithEvent,
            event: events::AUDIO_RECEIVED,
            session_id: "test-session".to_string(),
            payload: vec![1, 2, 3, 4, 5],
            ..Default::default()
        };

        let data = proto.marshal(&msg).unwrap();
        let decoded = proto.unmarshal(&data).unwrap();

        assert_eq!(decoded.msg_type, msg.msg_type);
        assert_eq!(decoded.flags, msg.flags);
        assert_eq!(decoded.event, msg.event);
        assert_eq!(decoded.session_id, msg.session_id);
        assert_eq!(decoded.payload, msg.payload);
    }
}
